### ISSUE TITLE
DX: Container.From() and Container.Build() creates fresh container

### ISSUE THREAD

**Issue by kpenfound** [dagger/dagger#4588]
Discord context: https://discord.com/channels/707636530424053791/1074736111131304057/1074747695765405696

Summary:
Container.From() and Container.Build() create a new container. Chaining this with `.client.Container()` instead of directly `client.From()`/`container.Build()`, or `client.ContainerFrom()`/`client.ContainerBuild()` introduces opportunity for potential footguns. 

For example:
```js
const foo = await client
      .container()
      .withEnvVariable('BUST_CACHE', Math.random().toString())
      .build(node)
      .withExec(['node', '--version'])
      .stdout();
```

The cache busting env variable will never actually make it to the engine since the environment is replaced (correct me if this is wrong)

**Comment by RonanQuigley** [comment]
👀 👋 I was the one that ran into this issue. 

In terms of DX, I just wanted to add that this was painful. Myself and a colleague at work lost an hour of our time trying to diagnose this. We then had to have the underlying behaviour explained to us over discord. 

I was thinking about this problem more, and I'd like to understand the thinking behind having an API design that starts off with an initial scratch container if `build` also produces a container for it to use?

**Comment by shykes** [comment]
> 👀 👋 I was the one that ran into this issue.
> 
> In terms of DX, I just wanted to add that this was painful. Myself and a colleague at work lost an hour of our time trying to diagnose this. We then had to have the underlying behaviour explained to us over discord.

Sorry about that...

> I was thinking about this problem more, and I'd like to understand the thinking behind having an API design that starts off with an initial scratch container if `build` also produces a container for it to use?

I dug out some context, see links below. The short version is that there were many constraints, and we tried our best to choose the least bad option. We were trying to reconcile 1) feature parity with the CUE API, 2) respecting the spirit and best practices of GraphQL as we understood it, 3) the ability to chain operations into single query, 4) our best guess at what the most pleasant DX might be. 

* [On `container { from }`](https://github.com/dagger/dagger/pull/3151#discussion_r981744599)
* [On `container { build }`](https://github.com/dagger/dagger/pull/3151#discussion_r981744599). Note: initially there was no `directory { dockerBuild }` (my fault :). We added it later.
* [On adding `directory { dockerBuild }`](https://github.com/dagger/dagger/issues/4015), featuring my Mea Culpa to @aluzzardi :)

**Comment by RonanQuigley** [comment]
@shykes Thanks. So by the looks of it, you're translating GraphQL's hierarchical query structure so that you can make type resolvers that can go infinitely deep?

I'm not too sure what the solution here is, or if there even is one. All I know was that it was tough to identify what was going wrong with the provided feedback. Also, having a chainable API implies that every subsequent call is building off of the previous call. 

At this stage, my two pence worth is that a better debugging experience would have helped, so once the DAG visualisation and the logging improvements are released I might have a better indication of what's going on under the hood. For the latter, it would be worth emphasising in some sort of debug level logging that the chains based off of container A will be wiped after starting container B, if that makes sense? This feels to me like it's working around the DX problem, however I do think better signposting is a start. 

As an example, if you're doing: 
 
 ```node ./dagger-pipeline --debug # or whatever cli dagger ends up with```
 
 You'd have logs like:
 
```
=== container A started 
... container A logs 
=== container A completed. All env vars reset # (or something that tracks the state applied to container A i.e. env vars etc)
```

**Comment by TomChv** [comment]
I'll work on this one asap :rocket:

**Comment by helderco** [comment]
In this case I'd rather return a clear error than trying to debug after the fact.

**Comment by vito** [comment]
Worth noting that this works fine if you use `From` and this is arguably an inconsistency between `From` and `Build`.

With `From` it changes out the filesystem but _preserves environment variables_: https://github.com/dagger/dagger/pull/4932

This was a follow-up change after the initial API design:

* https://github.com/dagger/dagger/issues/4085
* https://github.com/dagger/dagger/pull/4455

We're in an inconsistent state here, since we only fixed `From` while `Build` still replaces everything wholesale.

Maybe `Build` should just do the same thing?

**Comment by helderco** [comment]
That's interesting @vito, I wasn't aware of the use case in #4085.

I like that mental model. It makes sense, but we’re only merging `Env`. We should also merge `Label` and `ExposedPorts` for consistency, and do it in `Build` too. Also making sure this is properly documented in the API.

The `From` and `Build` names do feel like a replacement though, but I’m not sure it’s worth it to deprecate now. We should create an issue for that after this one is resolved anyway.

**Comment by helderco** [comment]
On merging, looking at the current implementation for `Env`:

https://github.com/dagger/dagger/blob/19447fd4cbc0069e408b6fdf08f924a0106b4a2b/core/container.go#L318

The old env var doesn’t *really* get replaced (because it’s an array of `NAME=VALUE` strings). For all intents and purposes it does, except that it’s still present in the image config:

```go
addr, err := client.Container().
    WithEnvVariable("PATH", "/usr/bin").
    From("alpine")
    Publish(ctx, "ttl.sh/xxx")
```

```shell
> docker inspect ...
    "Config": {
        "Env": [
            "PATH=/usr/bin",
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ],
    }
```

The same doesn’t happen with `Dockerfile`:

```Dockerfile
FROM alpine
ENV PATH=/bin
ENV PATH=/usr/bin
```

```shell
> docker inspect ...
    "Config": {
        "Env": [
            "PATH=/usr/bin"
        ],
    }
```

I know it’s nit picking, but I’d prefer that the old one was removed. 🙂

**Comment by mircubed** [comment]
@TomChv do you need any additional information before moving forward with this one?

**Comment by TomChv** [comment]
@mircubed No, I think I get how I can take care of the issue.

If I have correctly understand the problem, we would like to forward any container's configuration created pre `Build` or `From` to the new container.
That seems tricky to do but I'm working on it.

Or second solution is to explicitly adds debug message to inform the user that some configuration will be missing in the new container.

> The old env var doesn’t really get replaced (because it’s an array of NAME=VALUE strings). For all intents and purposes it does, except that it’s still present in the image config:

I totally agree! I'll work on that too

**Comment by helderco** [comment]
> If I have correctly understand the problem, we would like to forward any container's configuration created pre `Build` or `From` to the new container. 

Yeah, on `Build` and `From` just merge the new image config with the current image config instead of replacing it, but in a way that can be explained by comparing to the effect of using `.WithXXX`.
